### PR TITLE
Types search page

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
Started from pairing with @orta over Twitch: https://www.twitch.tv/videos/729447355 on these [Figma designs](https://www.figma.com/file/x8FJrNqj6oupqWn1s3uMg4/Website?node-id=1113%3A0).

**Initial notes (9/2)**: I decided to take Orta's advice and stop midway through implementing it, and get a good night's sleep instead 😄. But I'm excited about this page and plan on working on it more later this week/end if there's still work to be done. The major things left are:
- [x] Hook up typed search to the query param (https://github.com/microsoft/TypeSearch/pull/38)
- [x] ~Sending a _second_ search for the `@types/` / non-`@types` version of a module, depending on what the user searches, and turn the list of `@types/` packages into a `Set<string>` to inform the `definitelyTyped` prop of `HitResult`~ Looks like Algolia actually does have the types support data in its results? -> Yes! Resolved in 5d966e30.
- [x] Add the npm/yarn/pnpm selector
    - [x] Store it in localStorage
- [x] Format the results rows nicely 💅  _(and add the "Powered by Algolia" notice)_
    - [x] General layout/fonts
    - [x] Colors & image styling
    - [x] Time recency
    - [x] Mobile responsiveness
    - [x] Use a ~markdown~ quick regex parser to convert package descriptions to plaintext
- [x] Add an initial search on the page
    - [ ] Figure out what that initial search should be
- [x] General QA and miscellaneous resultant bug fixes
    - [x] Potential crash: querying for `reactrer` shows a `description` equal to `null` 

**Update 9/3** ([tweet](https://twitter.com/JoshuaKGoldberg/status/1301752211475759105)): Mostly done with the basic logic; still needs some touchups and quality polish.

**Update 9/4** Should be ready for review! I've played with it a bunch and am happy so far with it. The remaining questions I have are:
* What should initial homepage search query be?
* What should the "no results found" indicator be? _(read: is this silly placeholder logo variant the right kind of silly?)_

<table>
<thead>
<tr>
<th>State</th>
<th>Desktop</th>
<th>Tablet</th>
<th>Mobile</th>
</tr>
</thead>
<tbody>
<tr>
<th>Initial</th>
<td>Desktop screenshot of initial state</td>
<td>Tablet screenshot of initial state</td>
<td>Mobile screenshot of initial state</td>
</tr>
<tr>
<th>Results</th>
<td>Desktop screenshot of results state</td>
<td>Tablet screenshot of results state</td>
<td>Mobile screenshot of results state</td>
</tr>
<tr>
<th>No Results</th>
<td>Desktop screenshot of no results state</td>
<td>
Tablet screenshot of no results state</td>
<td>Mobile screenshot of no results state</td>
</tr>
</tbody>
</table>

Co-authored-by: Orta Therox <orta.therox+gh@gmail.com>

### CODE REVIEW

File: packages/typescriptlang-org/lib/bootup/ingestion/createRootPagesLocalized.ts [hunk 1]
```diff
@@ -44,6 +45,10 @@ export const createRootPagesLocalized = async (
       originalSitePath = originalSitePath.substring(0, originalSitePath.length - 3)
     }
 
+    // Always use /s for the path, because the join above in windows would be \
+    originalSitePath = originalSitePath.split("\\").join("/")
+    // fullpath = fullpath.split("\\").join("/");
```
Comment (JoshuaKGoldberg) [inline packages/typescriptlang-org/lib/bootup/ingestion/createRootPagesLocalized.ts]: Todo: would you like me to send these as a separate PR for Windows support?

File: package.json [hunk 2]
```diff
@@ -55,6 +56,7 @@
   },
   "dependencies": {
     "@types/express": "^4.17.6",
+    "classnames": "^2.2.6",
```
Comment (JoshuaKGoldberg) [inline package.json]: Is this an acceptable dependency to add? I'm a fan of using this cute little utility to make dynamic `className` generation a little smoother.

File: packages/typescriptlang-org/src/templates/pages/dt/search.tsx [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+                {"\n"}
+                &gt; {installer[0]} @types/${hit.name} {installer[1]}
+              </>
+            )}
+          </code>
+        </pre>
+      </td>
+    </tr>
+  )
+}
```
Comment (JoshuaKGoldberg) [inline packages/typescriptlang-org/src/templates/pages/dt/search.tsx]: This page is pretty huge and only getting bigger. Do you think it's time for moving some of these into a `components/` directory?

File: packages/typescriptlang-org/src/templates/pages/dt/types.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    host: string
+    path: string
+    project: string
+    url: string
+    user: string
+}
+
+export type SearchTypes = {
+    ts: boolean
+}
```
Comment (JoshuaKGoldberg) [inline packages/typescriptlang-org/src/templates/pages/dt/types.ts]: Todo: look for a place these are typed? I didn't find one on a cursory 30 second search.

Comment (JoshuaKGoldberg) [reply packages/typescriptlang-org/src/templates/pages/dt/types.ts]: Otherwise this is failing the CI build so I guess I'll move it to... somewhere 😄

File: package.json (hunk 2 above)
Comment (orta) [reply package.json]: For two callsites, probably not, given that it's like a single 60 LOC function, a trivial version could be added in the file, I've not quite got a utils folder, but maybe lib could do that (I agree that it's a clever abstraction, a nice declarative way to describe classes, here's a TS port of just the stuff you're using)

https://www.typescriptlang.org/play?#code/JAMwrgdgxgLglgewgAigDwBQDocEMBOA5gM4BcyGxM+cEhyAPspACYCmItbLjyIuAG2JsAlAG0AuiOQBvALAAoYMChIqqAbmLDiyALzJJAbkXIz5syAT4MqiOoL0EIZI+LT5Ci9-NwXGAEJHaTt4CDA2Ux9vKE1tNmIsAAcwYgALDGConwBfRWV8NhgwfBRYrR0sACsEWgwAImR6kWzkPIUou2IEATYsAQRCDFazdAakNnqAGj5BYWQAMgWmmAB3BGapkdRMeonp2aE2ReX6tY2Z6giTlbTCyZavZBEgA

File: packages/typescriptlang-org/src/components/search/ResultRow.tsx [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+import { cx } from "../../lib/cx"
+import { JoinedPackage } from "./types"
+
+export type ResultRowprops = {
+  installer: [string, string]
+  joinedPackage: JoinedPackage
+}
+
+export const ResultRow: React.FC<ResultRowprops> = ({
```
Comment (JoshuaKGoldberg) [inline packages/typescriptlang-org/src/components/search/ResultRow.tsx:16]: I honestly did try to keep things generally in the one file per the existing convention... but it got very big and I had a hard time keeping track of what's in it, so I gave into my inner craving and split it all out. 😞

File: packages/typescriptlang-org/src/components/search/joinSearchResults.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
+import { JoinedPackage, JoinedSearchResult, RawSearchResults, SearchHit } from "./types"
+
+export const joinSearchResults = (
+    [directResult, typedResult]: RawSearchResults['results']
```
Comment (JoshuaKGoldberg) [inline packages/typescriptlang-org/src/components/search/joinSearchResults.ts]: So, here's a bit of maybe "wasted" work: the first few packages I scanned through didn't happen to mention `types: { definitelytyped: "@types/name" }`, so I figured we'd need to send _two_ search requests... but on looking deeper, it looks like Algolia actually does have the right types data in results... so, assuming that data is filled out nicely, we don't actually need this whole joining-two-searches-together shenanigan. 

Pushing this commit now as a reference in case it's useful later. 🤷

File: packages/typescriptlang-org/src/templates/pages/dt/search.tsx (hunk 3 above)
Comment (JoshuaKGoldberg) [reply packages/typescriptlang-org/src/templates/pages/dt/search.tsx]: Update: I couldn't resist...

File: yarn.lock [hunk 7]
```diff
@@ -4070,14 +4070,14 @@ __metadata:
   languageName: node
   linkType: hard
 
-"@typescript/twoslash@0.6.2, @typescript/twoslash@workspace:packages/ts-twoslasher":
+"@typescript/twoslash@1.1.0, @typescript/twoslash@workspace:packages/ts-twoslasher":
```
Comment (JoshuaKGoldberg) [inline yarn.lock]: Todo: I'll `git checkout v2 -- yarn.lock` these...

File: packages/typescriptlang-org/src/components/search/ResultRow.tsx (hunk 5 above)
Comment (orta) [reply packages/typescriptlang-org/src/components/search/ResultRow.tsx:16]: Totally reasonable 👍🏻

File: packages/typescriptlang-org/lib/bootup/ingestion/createRootPagesLocalized.ts (hunk 1 above)
Comment (orta) [reply packages/typescriptlang-org/lib/bootup/ingestion/createRootPagesLocalized.ts]: Nah, if no-one's hit this issue (it only affects a few pages) it's fine

File: packages/typescriptlang-org/src/templates/pages/dt/search.tsx [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+const Index: React.FC<Props> = props => {
+  const [search, setSearch] = useState<string>(
+    new URLSearchParams(props.location.search).get("search") || ""
+  )
+  const result = useSearchResult(search)
+
+  useEffect(() => {
+    const params = new URLSearchParams(window.location.search)
+    params.set("search", search)
+    history.pushState(null, "", `${window.location.pathname}?${params}`)
```
Comment (orta) [inline packages/typescriptlang-org/src/templates/pages/dt/search.tsx]: win

**Review by orta — APPROVED** [review]
looking good so far

File: packages/typescriptlang-org/lib/bootup/ingestion/createRootPagesLocalized.ts (hunk 1 above)
Comment (JoshuaKGoldberg) [reply packages/typescriptlang-org/lib/bootup/ingestion/createRootPagesLocalized.ts]: Ok cool, filed https://github.com/microsoft/TypeScript-Website/issues/1043 for tracking. I'll remove these changes and try to remember to keep them just locally for my system.

File: packages/typescriptlang-org/src/components/search/ResultRow.tsx [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+          className={cx("typeImage", icon)}
+          role="img"
+        />
+      </td>
+      <td className="name">
+        <strong>{name}</strong>
+        {description}
+      </td>
+      <td className="updated">{modified}</td>
+      <td className="install">
```
Comment (JoshuaKGoldberg) [inline packages/typescriptlang-org/src/components/search/ResultRow.tsx:96]: Question: how would you like this to look on mobile? I'm guessing... hide the installation instructions, and on very small screens hide the "last updated"?

File: packages/typescriptlang-org/src/components/search/ResultRow.tsx [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+          aria-label={`Types ${label}`}
+          className={cx("typeImage", icon)}
+          role="img"
+        />
+      </td>
+      <td className="name">
+        <strong>{name}</strong>
+        {description}
+      </td>
+      <td className="updated">{modified}</td>
```
Comment (JoshuaKGoldberg) [inline packages/typescriptlang-org/src/components/search/ResultRow.tsx]: Question: do you have a preference for how to convert the updated timestamp to human text? My default would be `moment` from past usage, but it is a little large...

File: packages/typescriptlang-org/src/components/search/ResultRow.tsx [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+      <td className="via">
+        <div
+          aria-label={`Types ${label}`}
+          className={cx("typeImage", icon)}
+          role="img"
+        />
+      </td>
+      <td className="name">
+        <strong>{name}</strong>
+        {description}
```
Comment (JoshuaKGoldberg) [inline packages/typescriptlang-org/src/components/search/ResultRow.tsx]: Question: do you have a preference on method to convert this markdown description to plaintext? I would default to `marked` from past experience.

**Review by JoshuaKGoldberg — COMMENTED** [review]
I left some comments for you @orta on fixing up the last few bits. 🚀

File: packages/typescriptlang-org/src/components/search/ResultRow.tsx (hunk 10 above)
Comment (orta) [reply packages/typescriptlang-org/src/components/search/ResultRow.tsx]: Aye, use the platform, I use `Intl` in a few places now:

```ts
  const d = new Date(props.lastEdited);
  const dtf = new Intl.DateTimeFormat(
    props.lang,
    { year: "numeric", month: "short", day: "2-digit" },
  );
  const lastEdited = dtf.format(d);
```

```ts
  const releaseParts = intl.formatDateToParts(startDate, { month: "short", day: "numeric" })
```

File: packages/typescriptlang-org/src/components/search/ResultRow.tsx (hunk 11 above)
Comment (orta) [reply packages/typescriptlang-org/src/components/search/ResultRow.tsx]: Aye, no dependencies for this, I'm open to a rudimentary regex dropping `[xbc](yzx)` to `xbc` but even npm doesn't md parse

Screen Shot 2020-09-04 at 12 19 11 PM

File: packages/typescriptlang-org/src/lib/useLocalStorage.ts [hunk 12]
```diff
@@ -0,0 +1,12 @@
+import { useState } from "react"
+
+export const useLocalStorage = <State extends string>(key: string, initialValue: State) => {
+    const [value, setValue] = useState(globalThis.localStorage?.getItem(key) ?? initialValue)
```
Comment (orta) [inline packages/typescriptlang-org/src/lib/useLocalStorage.ts]: Careful, you have to [verify local storage](https://github.com/microsoft/TypeScript-website/blob/1c68e217fd03e5e02e6a7685457fca5a42b23f34/packages/typescriptlang-org/src/components/ShowExamples.tsx#L87-L95) exists first:

```
let hasLocalStorage = false
    try {
      hasLocalStorage = typeof localStorage !== `undefined`
    } catch (error) { }

    if (hasLocalStorage) {
      const examplesFromLS = localStorage.getItem("examples-seen") || "{}"
      seenExamples = JSON.parse(examplesFromLS)
    }
```

Otherwise usage will throw if someone has high enough privacy settings

File: packages/typescriptlang-org/src/components/search/ResultRow.tsx (hunk 9 above)
Comment (orta) [reply packages/typescriptlang-org/src/components/search/ResultRow.tsx:96]: Yep, I think just seeing the name and type and making the entire thing like to npm/DT is probably the right pattern - I'd imagine you're only doing this to prove that a dep has types in some format on mobile rather than doing anything with some depth

File: packages/typescriptlang-org/src/components/search/ResultRow.tsx (hunk 10 above)
Comment (JoshuaKGoldberg) [reply packages/typescriptlang-org/src/components/search/ResultRow.tsx]: Love it 😍

**Comment by orta** [comment]
haha, that no results, cloning and trying locally

**Comment by orta** [comment]
I've applied all my feedback inline directly (dark mode, some UX / copy changes), this is pretty much shippable right now.

I do have an unresolved question: What do you think the interaction pattern for linking to the source for the DTS should be? e.g. I'm looking at this:

Screen Shot 2020-09-05 at 5 25 03 PM
 
And I want to go to the async dep, should:

 - make just the title an anchor?
 - support clicking o the row in any section other than the code?
 - other?

**Comment by JoshuaKGoldberg** [comment]
Yeah that's a good question... how about this:

* DLs: link to npm
* Via: link to the GitHub source, as the GH repository for built-in and the DefinitelyTyped repository
* Module > name: link to npm

Everything else _feels_ like non-link plain text to me...

Additionally, as @danvk mentioned, it'd be great to have the install commands copy & pasteable, like on npmjs.com.

File: packages/typescriptlang-org/src/templates/pages/dt/search.tsx [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+  data: DTSearchPageQuery
+  location: Location
+  pageContext: any
+}
+
+const updateHistorySearch = debounce((search: string) => {
+  const params = new URLSearchParams(window.location.search)
+  params.set("search", search)
+  history.pushState(null, "", `${window.location.pathname}?${params}`)
+}, 250)
```
Comment (JoshuaKGoldberg) [inline packages/typescriptlang-org/src/templates/pages/dt/search.tsx:27]: This was a [fun little reminder](https://twitter.com/JoshuaKGoldberg/status/1302453905813536768) on how to properly investigate performance issues 😄. Calling this function repeatedly was slowing the page's render down.

File: packages/typescriptlang-org/src/components/search/ResultRow.tsx [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+  },
+}) => {
+  const npmUrl = `https://www.npmjs.com/package/${name}`
+  const [icon, label, viaUrl] =
+    types.ts === "included"
+      ? ["in", "included", repository?.url || npmUrl]
+      : [
+          "dt",
+          "from Definitely Typed",
+          `https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/${name}`,
```
Comment (JoshuaKGoldberg) [inline packages/typescriptlang-org/src/components/search/ResultRow.tsx]: Question: is this an assumption we can make? _(I haven't seen any indication otherwise)_

File: packages/typescriptlang-org/src/components/search/ResultRow.tsx [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+        ]
+
+  const initialCopyStatus = ["copy", `Copy ${name} installation script`]
```
Comment (JoshuaKGoldberg) [inline packages/typescriptlang-org/src/components/search/ResultRow.tsx]: Question: is there a better icon set we should use instead of `"copy"`/`"copied"`/`":("`? E.g. https://developer.microsoft.com/en-us/fluentui#/styles/web/icons, but just SVGs?

**Review by JoshuaKGoldberg — COMMENTED** [review]
Ok, from my perspective this is "ready to ship" - in that I have no more features I want to add, and only have questions on some choices made. 🎉

File: packages/typescriptlang-org/src/components/search/ResultRow.tsx (hunk 14 above)
Comment (orta) [reply packages/typescriptlang-org/src/components/search/ResultRow.tsx]: Close, it needs to handle `@absinthe/socket-apollo-link types` -> `absinthe__socket-apollo-link`

https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/absinthe__socket-apollo-link

**Comment by orta** [comment]
OK, I've given it an Intl pass, and handled my own on the DT link.

Re: copy, I'm alright with this technique, it conceptually playground links happen in the code samples 

Screen Shot 2020-09-06 at 10 04 49 AM

**Comment by orta** [comment]
Good to merge on green

**Comment by orta** [comment]
OK, cool, let's do this

**Comment by orta** [comment]
After https://github.com/microsoft/TypeScript-Website/actions/runs/241781529 completes, then https://www.staging-typescript.org/dt/search should work
